src/session/sync_session.rs

```diff
@@ -132,6 +132,51 @@ impl<P, S: Read + NonBlocking> Session<P, S> {
         }
     }
 
+    /// Expect waits until a pattern is matched. This continuously reads all the
+    /// available output, consumes it from the buffer and try to match on it
+    /// until a match is found.
+    ///
+    /// Typically, you'd want to use `expect_eager` over `expect` when the
+    /// command under test produces too much output to be processed lazily.
+    ///
+    /// If the method returns [Ok] it is guaranteed that at least 1 match was found.
+    ///
+    /// ```
+    /// let mut p = expectrl::spawn("echo 123").unwrap();
+    /// let m = p.expect_eager(expectrl::Regex("\\d+")).unwrap();
+    /// assert_eq!(m.first(), b"123");
+    /// ```
+    ///
+    /// It returns an error if timeout is reached.
+    /// You can specify a timeout value by [Session::set_expect_timeout] method.
+    pub fn expect_eager<E: Needle>(&mut self, expect: E) -> Result<Captures, Error> {
+        let start = time::Instant::now();
+        loop {
+            if let Some(timeout) = self.expect_timeout {
+                if start.elapsed() > timeout {
+                    return Err(Error::ExpectTimeout);
+                }
+            }
+
+            let eof = self.stream.read_available()?;
+            let data = self.stream.get_available();
+
+            let found = expect.check(data, eof)?;
+            if !found.is_empty() {
+                let end_index = Captures::right_most_index(&found);
+                let involved_bytes = data[..end_index].to_vec();
+                self.stream.consume_available(end_index);
+                return Ok(Captures::new(involved_bytes, found));
+            } else if !data.is_empty() {
+                let data_len = data.len();
+                self.stream.consume_available(data_len);
+                continue;
```

Comment:
But is this correct?

I mean what might happen essentially is we never find a match even though it was there.

Look at this example.

We have an output `Hello World`.
We are doing  `expect_eager("Hello World")`.
If `read_available` makes 2 reads to get the output.
First read will give `Hello` second `World`.
Because you clean the stream here the match will never come.